test/common/quic/envoy_quic_session_cache_test.cc

```diff
@@ -0,0 +1,343 @@
+        SSL_SESSION_from_bytes(reinterpret_cast<const uint8_t*>(cached_session.data()),
+                               cached_session.size(), ssl_ctx_.get());
+    SSL_SESSION_set_time(session, std::chrono::system_clock::to_time_t(time_source_.systemTime()));
+    SSL_SESSION_set_timeout(session, timeout);
+    return bssl::UniquePtr<SSL_SESSION>(session);
+  }
+
+  bssl::UniquePtr<SSL_CTX> ssl_ctx_;
+  FakeTimeSource time_source_;
+  EnvoyQuicSessionCache cache_;
+};
+
+// Tests that simple insertion and lookup work correctly.
+TEST_F(EnvoyQuicSessionCacheTest, SingleSession) {
+  std::unique_ptr<quic::TransportParameters> params = makeFakeTransportParams();
+  bssl::UniquePtr<SSL_SESSION> session = makeSession();
+  quic::QuicServerId id1("a.com", 443);
+
+  std::unique_ptr<quic::TransportParameters> params2 = makeFakeTransportParams();
+  bssl::UniquePtr<SSL_SESSION> session2 = makeSession();
+  SSL_SESSION* unowned2 = session2.get();
+  quic::QuicServerId id2("b.com", 443);
+
+  EXPECT_EQ(nullptr, cache_.Lookup(id1, ssl_ctx_.get()));
+  EXPECT_EQ(nullptr, cache_.Lookup(id2, ssl_ctx_.get()));
+  EXPECT_EQ(0u, cache_.size());
+
+  cache_.Insert(id1, std::move(session), *params, nullptr);
+  EXPECT_EQ(1u, cache_.size());
+  std::unique_ptr<quic::QuicResumptionState> resumption_state = cache_.Lookup(id1, ssl_ctx_.get());
+  ASSERT_NE(resumption_state, nullptr);
+  ASSERT_NE(resumption_state->transport_params, nullptr);
+  EXPECT_EQ(*params, *(resumption_state->transport_params));
+  EXPECT_EQ(nullptr, cache_.Lookup(id2, ssl_ctx_.get()));
+  // No session is available for id1, even though the entry exists.
+  EXPECT_EQ(1u, cache_.size());
+  EXPECT_EQ(nullptr, cache_.Lookup(id1, ssl_ctx_.get()));
+  // Lookup() will trigger a deletion of invalid entry.
+  EXPECT_EQ(0u, cache_.size());
+
+  bssl::UniquePtr<SSL_SESSION> session3 = makeSession();
+  SSL_SESSION* unowned3 = session3.get();
+  quic::QuicServerId id3("c.com", 443);
+  cache_.Insert(id3, std::move(session3), *params, nullptr);
+  cache_.Insert(id2, std::move(session2), *params2, nullptr);
+  EXPECT_EQ(2u, cache_.size());
+  std::unique_ptr<quic::QuicResumptionState> resumption_state2 = cache_.Lookup(id2, ssl_ctx_.get());
+  ASSERT_NE(resumption_state2, nullptr);
+  EXPECT_EQ(unowned2, resumption_state2->tls_session.get());
+  std::unique_ptr<quic::QuicResumptionState> resumption_state3 = cache_.Lookup(id3, ssl_ctx_.get());
+  ASSERT_NE(resumption_state3, nullptr);
+  EXPECT_EQ(unowned3, resumption_state3->tls_session.get());
+
+  // Verify that the cache is cleared after Lookups.
+  EXPECT_EQ(nullptr, cache_.Lookup(id1, ssl_ctx_.get()));
+  EXPECT_EQ(nullptr, cache_.Lookup(id2, ssl_ctx_.get()));
+  EXPECT_EQ(nullptr, cache_.Lookup(id3, ssl_ctx_.get()));
+  EXPECT_EQ(0u, cache_.size());
+}
+
+TEST_F(EnvoyQuicSessionCacheTest, MultipleSessions) {
+  std::unique_ptr<quic::TransportParameters> params = makeFakeTransportParams();
+  bssl::UniquePtr<SSL_SESSION> session = makeSession();
```

Comment:
I know they're not used in all tests, but would it make sense to make some of these member variables just to cut down on test set-up?

Reply:
I moved params to a member but not session because the session ownership is transferred to the cache